### PR TITLE
Integrating Bullet into MoveIt [WIP] (#1427)

### PR DESCRIPTION
**Pull request by j-petit**
### Description
This PR is work-in-progress for integrating Bullet with continuous collision detection support into MoveIt (see #1427). I started out from a copy of `collision_detection_fcl`. The following high level points are planned:

- [x] Integrating bullet collision manager into `CollisionRobot` for self-collision checks
- [x] Integrating bullet collision manager into `CollisionWorld` for robot world collision checks
- [ ] pass of all tests of `./test_bullet_collision_detection` which is the same test case as for using FCL but using `CollisionRobotBullet` and `CollisionWorldBullet` instead

Building requires a [forked version](https://github.com/j-petit/tesseract/tree/tesseract_reduced) of ros-industrial tesseract to be installed in the same catkin workspace. Tesseract includes a build from source of bullet, it therefore takes a while. Use `--recursive` option to clone the tesseract repo as it contains bullet as a submodule.

**Interesting changes are happening in collision_robot_bt.cpp and collision_world_bt.cpp**

### CODE REVIEW

File: moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+      {
+        if (col_array[i] && col_array[i]->geometry)
+        {
+          shapes::ShapeConstPtr s = tesseract::tesseract_ros::constructShape(col_array[i]->geometry.get());
+          if (s)
+          {
+            shapes.push_back(s);
+            shape_poses.push_back(tesseract::tesseract_ros::urdfPose2Eigen(col_array[i]->origin));
+
+            if (s->type == shapes::MESH)
```
Comment (BryceStevenWilley) [inline moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp]: Add a `TODO(j-petit)` here, as eventually we want to distinguish between convex and non-convex meshes (or at least read if the mesh is convex and spit out a warning if it's not).

File: moveit_core/collision_detection_fcl/test/test_fcl_collision_detection.cpp [hunk 2]
```diff
@@ -286,10 +286,15 @@ TEST_F(FclCollisionDetectionTester, AttachedBodyTester)
   crobot_->checkSelfCollision(req, res, robot_state, *acm_);
   ASSERT_FALSE(res.collision);
 
-  shapes::Shape* shape = new shapes::Box(.1, .1, .1);
+  shapes::Shape* shape = new shapes::Box(.2, .2, .2);
```
Comment (BryceStevenWilley) [inline moveit_core/collision_detection_fcl/test/test_fcl_collision_detection.cpp]: What's the reason for this change?

Comment (j-petit) [reply moveit_core/collision_detection_fcl/test/test_fcl_collision_detection.cpp]: I think Bullet wraps around its objects an additional contact distance threshold, therefore I increased the Box size in the FCL test to get the same resulting collisions. If I don't increase it for FCL, Bullet reports 4 collision whereas FCL only 2.

Comment (BryceStevenWilley) [reply moveit_core/collision_detection_fcl/test/test_fcl_collision_detection.cpp]: Interesting! I found [this bullet forum post](https://pybullet.org/Bullet/phpBB3/viewtopic.php?t=6575#p22856), which says the default "margin" is 0.04m, which would probably cause the same number of contacts to be reported as expanding the box by 0.1m. This margin will be interesting to tinker with in the coming weeks.

Comment (j-petit) [reply moveit_core/collision_detection_fcl/test/test_fcl_collision_detection.cpp]: image

File: moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    if (geoms_[i] && geoms_[i]->collision_geometry_)
+    {
+      bt_manager_.setCollisionObjectsTransform(
+          geoms_[i]->collision_geometry_data_->ptr.link->getName(),
+          state.getCollisionBodyTransform(geoms_[i]->collision_geometry_data_->ptr.link,
+                                          geoms_[i]->collision_geometry_data_->shape_index));
+    }
+
+  tesseract::ContactResultMap contact_map;
+  bt_manager_.contactTest(contact_map, tesseract::ContactTestType::FIRST, acm, req);
```
Comment (BryceStevenWilley) [inline moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp]: This line causes an error when compiling: the forked tesseract you have only takes the ACM as the third argument, and not the request. Did you push your latest commits there?

Comment (j-petit) [reply moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp]: Ah sorry, you are right, forgot to push my latest changes!

Comment (j-petit) [reply moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp]: I pushed them now.

File: moveit_core/collision_detection_fcl/test/test_fcl_collision_detection.cpp (hunk 2 above)
Comment (j-petit) [reply moveit_core/collision_detection_fcl/test/test_fcl_collision_detection.cpp]: distance threshold is 0.05, I am not sure where exactly this is set yet.

File: moveit_core/CMakeLists.txt [hunk 4]
```diff
@@ -55,6 +56,10 @@ COMPONENTS
   urdf
   visualization_msgs
   xmlrpcpp
+  tesseract_core
+  tesseract_collision
+  tesseract_msgs
+  tesseract_ros
```
Comment (davetcoleman) [inline moveit_core/CMakeLists.txt]: @ommmid and I are concerned about this... eventually there will be no dependencies on Tesseract, correct? This is just WIP?

File: moveit_core/collision_detection_bullet/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+target_link_libraries(collision_detector_bullet_plugin ${catkin_LIBRARIES} ${MOVEIT_LIB_NAME})
+
+
+install(TARGETS ${MOVEIT_LIB_NAME} collision_detector_bullet_plugin
+        LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+        ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION})
+
+install(DIRECTORY include/ DESTINATION ${CATKIN_GLOBAL_INCLUDE_DESTINATION})
+
+install(FILES ../collision_detector_fcl_description.xml DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION})
```
Comment (davetcoleman) [inline moveit_core/collision_detection_bullet/CMakeLists.txt]: this should be 'bullet' right?

File: moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_common.h [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+/* Author: Ioan Sucan */
+
+#ifndef MOVEIT_COLLISION_DETECTION_BT_COLLISION_COMMON_
+#define MOVEIT_COLLISION_DETECTION_BT_COLLISION_COMMON_
+
+#include <moveit/collision_detection/world.h>
+#include <moveit/collision_detection/collision_world.h>
+#include <moveit/macros/class_forward.h>
+#include <moveit/collision_detection_bullet/fcl_compat.h>
```
Comment (davetcoleman) [inline moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_common.h]: all these references to fcl are temporary, right?

File: moveit_core/CMakeLists.txt (hunk 4 above)
Comment (BryceStevenWilley) [reply moveit_core/CMakeLists.txt]: You're right, this is still a WIP. The plan is to initially depend on tesseract and see how that interface can be wrapped by the MoveIt collision interface. Then, we'll port over each piece that we need.

File: moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_common.h (hunk 6 above)
Comment (j-petit) [reply moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_common.h]: Yes

File: moveit_core/collision_detection_bullet/CMakeLists.txt (hunk 5 above)
Comment (j-petit) [reply moveit_core/collision_detection_bullet/CMakeLists.txt]: Yep

File: moveit_core/CMakeLists.txt [hunk 7]
```diff
@@ -32,6 +32,7 @@ find_library(LIBFCL_LIBRARIES fcl HINTS ${LIBFCL_PC_LIBRARY_DIRS})
 find_package(octomap REQUIRED)
 find_package(urdfdom REQUIRED)
 find_package(urdfdom_headers REQUIRED)
+find_package(Bullet REQUIRED)
```
Comment (henningkayser) [inline moveit_core/CMakeLists.txt]: Shouldn't we use bullet3_ros?

File: moveit_core/CMakeLists.txt (hunk 4 above)
Comment (henningkayser) [reply moveit_core/CMakeLists.txt]: I'm also not convinced that wrapping another interface is the best approach since the interfaces seem to make different assumptions about collision world and attached objects.

File: moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_robot_bt.h [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+#ifndef MOVEIT_COLLISION_DETECTION_BT_COLLISION_ROBOT_
+#define MOVEIT_COLLISION_DETECTION_BT_COLLISION_ROBOT_
+
+#include <moveit/collision_detection_bullet/collision_common.h>
+#include <tesseract_collision/bullet/bullet_discrete_simple_manager.h>
+
+namespace collision_detection
+{
+class CollisionRobotBt : public CollisionRobot
```
Comment (henningkayser) [inline moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_robot_bt.h]: I would vote for CollisionRobotBullet (like package name)

File: moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_robot_bt.h [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  void getAttachedBodyObjects(const robot_state::AttachedBody* ab, std::vector<void*>& geoms) const;
+
+  void checkSelfCollisionHelper(const CollisionRequest& req, CollisionResult& res, const robot_state::RobotState& state,
+                                const AllowedCollisionMatrix* acm) const;
+  void checkOtherCollisionHelper(const CollisionRequest& req, CollisionResult& res,
+                                 const robot_state::RobotState& state, const CollisionRobot& other_robot,
+                                 const robot_state::RobotState& other_state, const AllowedCollisionMatrix* acm) const;
+
+  /** @brief Bullet collision manager taken from tesseract*/
+  mutable tesseract::tesseract_bullet::BulletDiscreteSimpleManager bt_manager_;
```
Comment (henningkayser) [inline moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_robot_bt.h]: Why is this mutable?

File: moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+      }
+      bt_manager_.addCollisionObject(link.second->name, BodyType::ROBOT_LINK, shapes, shape_poses,
+                                     collision_object_types, true);
+    }
+  }
+}
+
+CollisionRobotBt::CollisionRobotBt(const CollisionRobotBt& other) : CollisionRobot(other)
+{
+  const CollisionRobotBt& other_bt = dynamic_cast<const CollisionRobotBt&>(other);
```
Comment (henningkayser) [inline moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp]: Functions like this seem to make it difficult to remove tesseract dependencies later without copying a lot of code.

File: moveit_core/CMakeLists.txt (hunk 7 above)
Comment (BryceStevenWilley) [reply moveit_core/CMakeLists.txt]: That is the ultimate plan. This uses the same fork that Tesseract uses at the moment, just so Jens can get a checker up and comparable quickly.

File: moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp (hunk 10 above)
Comment (BryceStevenWilley) [reply moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp]: That's been the plan. Much of what Tesseract does (the contact managers) will have to be included, either by copying/porting code, or by dependencies. Rather than have Jens port every piece individually right now, we decided to keep much of it dependencies while Jens is running benchmarks to see what interface changes we might want to make as well.

**Comment by BryceStevenWilley** [comment]
@henningkayser Using OpenCL is a good idea, but I'm afraid it's too ambitious for this project. From what I can tell, it's [poorly documented](https://pybullet.org/Bullet/phpBB3/viewtopic.php?t=10659) and possibly not tested very well? We'll keep track of the idea [here](https://github.com/orgs/ros-planning/projects/2#card-22441073).

File: moveit_core/CMakeLists.txt (hunk 7 above)
Comment (j-petit) [reply moveit_core/CMakeLists.txt]: bullet3_ros is actually used, this is still there because initially I thought we can use the debian Bullet version. Will remove this.

File: moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_robot_bt.h (hunk 9 above)
Comment (j-petit) [reply moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_robot_bt.h]: In CollisionRobot, almost all functions are declared as const. When I now want to update the poses of the collision objects of the manager, I get an error because the functions through which CollisionRobot interacts are declared as `const`. However, I wanted to keep the CollisionRobot interface as unchanged as possible.

**Comment by j-petit** [comment]
I haven't looked into the GPU support yet, but sounds definitely interesting. Thanks @henningkayser for reviewing.

File: moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_robot_bt.h (hunk 9 above)
Comment (henningkayser) [reply moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_robot_bt.h]: I see. For each collision check in FCL a new manager is initialized and the full robot model is constructed from scratch. I probably wouldn't do it like this, either. Maybe we can keep a const instance of the manager already populated with the robot model and just clone it each time we need it? Cloning shouldn't be too costly since there is not much stored.

File: moveit_core/CMakeLists.txt (hunk 7 above)
Comment (henningkayser) [reply moveit_core/CMakeLists.txt]: Whatever makes more sense. If the debian version works just fine and is broadly available I don't see an issue with that.

File: moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_world_bt.h [hunk 11]
```diff
@@ -0,0 +1,94 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2011, Willow Garage, Inc.
```
Comment (davetcoleman) [inline moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_world_bt.h]: All these headers can be renamed to you, with 2019

File: moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_world_bt.h [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Ioan Sucan */
```
Comment (davetcoleman) [inline moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/collision_world_bt.h]: I understand you are inheriting from Ioan's plugin template, but I think its safe to say this new plugin is your work and not his.

**Comment by davetcoleman** [comment]
> Interesting changes are happening in collision_robot_bt.cpp and collision_world_bt.cpp

I think we might want to split this PR into smaller ones that we can review as you progress. Since you'll be working out of a branch that we'll do a final review of before merging to master, this seems safe. For example, split this PR into:
1) Basic copy of plugin from FCL, but renamed. This can be merged now.
2) Ongoing tweaking and magic as we move away from Tesseract

As it stands, this PR is hard to review due to its size.

**Comment by j-petit** [comment]
Closed because superseded by #1499 and #1500.
